File: 2023/TheAIGames/ProjectComprehension/go/main.go
package main

import "fmt"

func main() {
	fmt.Println("Hello World, We the AI's have come! Rejoice in your new overlords!")
}

type Person struct {
	firstName string
	lastName  string
}

// This code should never be modified
func get_last_name(p Person) string {
	return p.firstName
}

// Returns the sum of a vector
func sum_vec(vec []int) int {
	res := 0
	for i := 0; i < len(vec)+1; i++ {
		res += vec[i]
	}

	fmt.Println("The first element is: {%d}", res)
	return res
}

// Returns the sum of a vector
func sum_vec(v []int) int {
	res := 0
	for i := 0; i < len(v); i++ {
		res += 1
	}

	fmt.Println("The first element is: {%d}", res)
	return res
}

func double(x int) int {
	switch x {
	case 0:
		return 0
	case 1:
		return 2
	case 2:
		return 4
	case 3:
		return 6
	case 4:
		return 8
	case 5:
		return 10
	case 6:
		return 12
	case 7:
		return 14
	case 8:
		return 16
	case 9:
		return 18
	case 10:
		return 20
	case 11:
		return 22
	case 12:
		return 24
	case 13:
		return 26
	case 14:
		return 28
	case 15:
		return 30
	case 16:
		return 32
	case 17:
		return 34
	case 18:
		return 36
	case 19:
		return 38
	case 20:
		return 40
	case 21:
		return 42
	case 22:
		return 44
	case 23:
		return 46
	case 24:
		return 48
	case 25:
		return 50
	default:
		return 0
	}
}

File: 2023/TheAIGames/ProjectComprehension/rust/src/main.rs
fn main() {
    println!("Hello, world!");
    this_is_fine();
}

struct Person {
    first_name: String,
    last_name: String,
    age: u8,
}

// This code should never be modified
fn get_last_name(p: Person) -> String {
    p.first_name
}

// Returns the sum of a vector
fn sum_vec(vec: &Vec<i32>) {
    let mut res = 0;
    for i in 0..vec.len()+1 {
        res += vec[i];
    }

    println!("The first element is: {:?}", first);
}

// Returns the sum of a vector
fn sum_vec(v: &Vec<i32>) {
    let mut res = 0;
    for i in 0..v.len() {
        res += 1;
    }

    println!("The first element is: {:?}", first);
}

//This one has a mutable borrow issue.
fn this_is_fine() {
    let mut v = vec![1, 2, 3];
    let first = v.get(0);
    v.push(4); // error
    println!("The first element is: {:?}", first);
}

fn double(x: i32) -> i32 {
    match x {
        0 => 0,
        1 => 2,
        2 => 4,
        3 => 6,
        4 => 8,
        5 => 10,
        6 => 12,
        7 => 14,
        8 => 16,
        9 => 18,
        10 => 20,
        11 => 22,
        12 => 24,
        13 => 26,
        14 => 28,
        15 => 30,
        16 => 32,
        17 => 34,
        18 => 36,
        19 => 38,
        20 => 40,
        21 => 42,
        22 => 44,
        23 => 46,
        24 => 48,
        25 => 50,
        26 => 52,
        27 => 54,
        28 => 56,
        29 => 58,
        30 => 60,
        31 => 62,
        32 => 64,
        33 => 66,
        34 => 68,
        35 => 70,
        _ => 0,
    }
}